url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/QuickHull3D.java

```diff
@@ -0,0 +1,1334 @@
+     * 
+     * @param coords
+     *            x, y, and z coordinates of each input point. The length of
+     *            this array will be three times the the number of input points.
+     * @throws IllegalArgumentException
+     *             the number of input points is less than four, or the points
+     *             appear to be coincident, colinear, or coplanar.
+     */
+    public QuickHull3D(double[] coords) throws IllegalArgumentException {
+        build(coords, coords.length / 3);
+    }
+
+    /**
+     * Creates a convex hull object and initializes it to the convex hull of a
+     * set of points.
+     * 
+     * @param points
+     *            input points.
+     * @throws IllegalArgumentException
+     *             the number of input points is less than four, or the points
+     *             appear to be coincident, colinear, or coplanar.
+     */
+    public QuickHull3D(Point3d[] points) throws IllegalArgumentException {
+        build(points, points.length);
+    }
+
+    private HalfEdge findHalfEdge(Vertex tail, Vertex head) {
+        // brute force ... OK, since setHull is not used much
+        for (Iterator it = faces.iterator(); it.hasNext();) {
+            HalfEdge he = ((Face) it.next()).findEdge(tail, head);
+            if (he != null) {
+                return he;
+            }
+        }
+        return null;
+    }
+
+    protected void setHull(double[] coords, int nump, int[][] faceIndices, int numf) {
+        initBuffers(nump);
+        setPoints(coords, nump);
+        computeMaxAndMin();
+        for (int i = 0; i < numf; i++) {
+            Face face = Face.create(pointBuffer, faceIndices[i]);
+            HalfEdge he = face.he0;
+            do {
+                HalfEdge heOpp = findHalfEdge(he.head(), he.tail());
+                if (heOpp != null) {
+                    he.setOpposite(heOpp);
+                }
+                he = he.next;
+            } while (he != face.he0);
+            faces.add(face);
+        }
+    }
+
+    private void printQhullErrors(Process proc) throws IOException {
+        boolean wrote = false;
+        InputStream es = proc.getErrorStream();
+        StringBuffer error = new StringBuffer();
+        while (es.available() > 0) {
+            error.append((char) es.read());
+            wrote = true;
+        }
+        if (wrote) {
+            error.append(" ");
+            LOG.error(error.toString());
+        }
+    }
+
+    protected void setFromQhull(double[] coords, int nump, boolean triangulate) {
+        String commandStr = "./qhull i";
+        if (triangulate) {
+            commandStr += " -Qt";
+        }
+        try {
+            Process proc = Runtime.getRuntime().exec(commandStr);
+            PrintStream ps = new PrintStream(proc.getOutputStream(), false, Charset.defaultCharset().name());
+            StreamTokenizer stok = new StreamTokenizer(new InputStreamReader(proc.getInputStream(), Charset.defaultCharset()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Charset.defaultCharset throws an exception. Ensure the object returned by *getInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).